### ISSUE TITLE
Arduino Experts:  Question about breaking Marlin up into modules

### ISSUE THREAD

**Issue by Roxy-3DPrintBoard** [MarlinFirmware/MarlinDev#395]
This is a question to the Arduino Experts:

Does the ability to have a sketch be comprised of separate folders allow us to break up the code into separate modules.  In the past, we had to have everything in the same directory, and in fact, currently some of the files (like Marlin_main.cpp) have kept getting bigger and bigger because we can't seem to break them up.

We started down a path of turning Marlin into a library just so we could start breaking it up into modules.  But there are a lot of limitations to going that path.  First, you can only place the sketch in C:\Users\name\Documents\Arduino and even worse, it won't work if there are other sketches there.

The release notes seem very incomplete and I can't find information within the Arduino project on how to even use the multiple folders that it supposedly supports.

Can somebody jump in and enlighten us?

**Comment by MagoKimbra** [comment]
If you bring everything back as it was before in a sketch, now with Arduino IDE 1.6.7 you can include files in folders and sub-folders ...
Example:
#include "module/motion/stepper.h"
#Include "module/language/english/language.h"

**Comment by Roxy-3DPrintBoard** [comment]
I don't think we have a need to do it...  Can we use paths that are up above and outside of the folder that the sketch is in?

**Comment by MagoKimbra** [comment]
Exactly ... If you look at my fork is done right so ...

**Comment by thinkyhead** [comment]
I want very much to organize in subfolders, and the MagoKimbra layout is a good model, but I felt that Marlin 1.1.0 should be compatible with older versions of Arduino and other IDE systems that don't have the 1.6.7 toolset yet. Basically, because the community deserves a really good release following the old paradigm before we force them to move to a new one.

Would it make sense to (1) call Marlin with the new layout 1.1.1, or 1.2.0, and (2) should we curate a version of Marlin that uses the older layout for the convenience of users who prefer or are stuck with older toolsets?

**Comment by Roxy-3DPrintBoard** [comment]
> I want very much to organize in subfolders, and the MagoKimbra layout is a good model, but I felt that Marlin 1.1.0 should be compatible with older versions of Arduino and other IDE systems that don't have the 1.6.7 toolset yet. Basically, because the community deserves a really good release following the old paradigm before we force them to move to a new one.

@thinkyhead    When we get done with the Release Candidates, we should take a look at how hard it would be to merge all of the bug fixes in it to the file structure that MagoKimbra has.   It is probably too much to hope for, but wouldn't it be nice to have the best of both worlds???

**Comment by jbrazio** [comment]
As Marlin is dependent on Arduino as it's toolkit, if the toolkit evolves I do not believe the burden of curating legacy Marlin code should fall into core team hands. But I do agree with @thinkyhead were a transition period should be somehow in place in order not to unleash user rage. A stable release shall be then marked as the final one supporting Arduino toolkit 1.x and the team must move forward following mainstream.

**Comment by Roxy-3DPrintBoard** [comment]
> As Marlin is dependent on Arduino as it's toolkit, if the toolkit evolves I do not believe the burden of curating legacy Marlin code should fall into core team hands. But I do agree with @thinkyhead were a transition period should be somehow in place in order not to unleash user rage. A stable release shall be then marked as the final one supporting Arduino toolkit 1.x and the team must move forward following mainstream.

I'm not sure I understand what you are trying to say.   But here is what I'm thinking...   

If we can start moving functionality into separate folders it makes it much easier to start cleaning up the code.   We would still be using Arduino, and to the user the only thing that changed is the organization of the files when they unzip the .ZIP file to build the sketch.    

In fact, I think we would still try to have a configuration.h file that controlled the bulk or high level of the configuration.   Where we might want to have discussions is on topics like Auto Bed Leveling where there are so many sub-features that need to be tuned.   It may make sense to have reasonable defaults but bury those down in the folder where they get used.   And if you need to tweak a setting from its (reasonable) default, you would not find that in the main configuration.h file.   But like I say, that is just my thinking of how we might want to organize things assuming we can start breaking up the code into modules.

If we went this route, that is probably very easy for the user.  When they upgrade to a version of the firmware that is organized around the new file layout, they would do what they always do.  They would do a diff on their configuration.h file and compare it to the new configuration.h file to see if there are any new settings they need to cross over.    And then they compile and load it...

I don't think trying to cross our RCBugFix's over to the file layout that MagoKimbra has will cause issues for the users.

**Comment by jbrazio** [comment]
Maybe I got it wrong, but if we change the layout then the src code is incompatible with previous versions of Arduino right ? This was the context of my answer but once again I may have got the wrong understanding.

**Comment by jbrazio** [comment]
Overall I agree with your approach Roxy.

**Comment by Roxy-3DPrintBoard** [comment]
It complicates things a little for people that maintain and do things in their own branch.  But that statement is true of any change we make.   I think what we would do is try to make the minimal number of changes to slice the code apart.    That would make it much easier for somebody to cross their changes over to the new code base.   And it is from this foundation that we would try to continue the development work.

I think the topic deserves discussion to make sure we hear all of the concerns and get the best thinking out on the table.    After the discussion we will do our best to minimize the impact on people trying to maintain their own forks.   But we really do need to start cleaning up the code.   Marlin_main is over 5000 lines long just because we could not slice it up and use the current tool set.

**Comment by gddeen** [comment]
A user like me goes in and picks up a Marlin build. 
Their Concerns: Does it work with my printer and have all previous features.
They change Configuration.h.  Maybe make a few dozen changes elsewhere.
They learn quickly that it is difficult to keep up and just keep a configuration.h
and throw away all their other changes.

What they want is to select a Marlin release. merge dev "features 1, 2, 37" into that release.
Merge their previous configuration.h changes onto the release with the features (and save that delta). When a new base with the features 1, 2, 37 are merged into the release they want to be notified.

At that time, repeat...

1,000 fixes makes a user throw away their changes the same as breaking up files.

**Comment by Roxy-3DPrintBoard** [comment]
I said:

> If we went this route, that is probably very easy for the user. When they upgrade to a version of the firmware that is organized around the new file layout, they would do what they always do. They would do a diff on their configuration.h file and compare it to the new configuration.h file to see if there are any new settings they need to cross over. 

and:

> I think what we would do is try to make the minimal number of changes to slice the code apart. That would make it much easier for somebody to cross their changes over to the new code base. And it is from this foundation that we would try to continue the development work.

What part of trying to make it as easy as possible for the user don't you understand?

But here is something for YOU to understand:  The developers here literally have put in 1000's of hours of effort to design, write, debug and maintain the code.  The code base desperately needs to be cleaned up so that it can be reliable and easier to continue development.   If you don't want that, you can just stick with an old version you know and love.   If you want the code base to evolve, you are going to have to spend an extra 15 or 20 minutes when you switch to the latest and greatest version the first time.     

If we are going to make the code better, there have to be changes made to it.  It isn't realistic to say "I want to the code to be better but you are not allowed to make any changes."

**Comment by gddeen** [comment]
Yes Ma'am,
> I said:

> If we went this route, that is probably very easy for the user. When they upgrade to a version of the firmware that is organized around the new file layout, they would do what they always do. They would do a diff on their configuration.h file and compare it to the new configuration.h file to see if there are any new settings they need to cross over.
and:

> think what we would do is try to make the minimal number of changes to slice the code apart. That would make it much easier for somebody to cross their changes over to the new code base. And it is from this foundation that we would try to continue the development work. What part of trying to make it as easy as possible for the user don't you understand?

#Serious response to your questions
Why do you care about single file with 30,000 lines or 10,000 files. With github creating items for each change they "logically" are identical.

They practically aren't identical because of the (I)DEs used during the development process by the different roles of people. So, I outlined a role. Here is a "typical 3dprinter owner firmware patcher." They use the arduino software to build. They download from github. They gave up changing because Arduino lacks the tools to allow easy merge. Outline how to use merge tools might make their job seconds, not longer.

#Reason I Resonded
I had expected in response: "Here is a typical Marlin developer role. We use github desktop for new pulls, tracking issues. We have object oriented development platforms which edit, build, search, cross reference with new files for everything.  The majority use...'kjkjhifif'"**

I had not expected: "Jane you ignorant slut. You wouldn't know a reentrant call by reference if one probed your bits! This is hard work, sit back and let the intelligent people (that are coding on 8 bit processors) work." (I hope you laughed...)
 
Evolution and extra time usually are contraindicated.

My goal was to look at all of the roles represented and see if there are tools, changes to move everyone along. You did ask for this. Or, that's how I understood your original question.

#Elaborations
So far, because that single class is only changing 1 file, you could slice and dice every line in every file.
Scripts could be used to preprocess and retrieve configuration settings into implementation descriptions. IE, you wouldn't have to include the configuration.h file. IDE platforms could be targeted. Older platforms could be cast in stone, etc.

I am not sure of all of the platforms supported, CNC machines, coke machines, arms, printers. Developers, Retail rework, high school kids, old farts, etc. The scope of the effort I heard was make it nifty for mid-level firmware developers.

Also, I don't consider the current code very difficult at all. Marlin is a big file but it is broken into 3 logical sections.  An 'OK' editor removes with multiple views, functional searching, context sensitive support makes it easy.  

Build difficulty. Because Arduino handles that, the load module is kinda guessed at. I assume multiple files are wanted to allow multiple platforms to compile along with the traditional file structures and not uniquely C/C++. I.E. support for graphical dev/build platforms for the displays, traditional networking platforms, and mathematical object platforms.

Direction strategy. Maybe you wish to co-exist, merge into, pull from Arduino, iOS, Android, unix, web apps? I dunno.

So, if anyone is voting... My vote is for me to sit back and follow. I would hope Marlin defines state of the art IDEs to be used and how each player in the game would join in. Hackers, dabblers, full blooded programmers (like me), marketers, manufacturers, students. I would vote to define the set of equipment that would want to use the Marlin firmware and its purpose of that equipment. I would vote for descriptions of the work/effort so people could engage and profit in Marlin.

**But, I am a happy go lucky intellectual optimist. I enjoy other's perspectives.**

**Comment by Roxy-3DPrintBoard** [comment]
I apologize.   I took your message to mean you had some changes you added to your fork and you were not willing to spend a few minutes to move to an updated code base.    

It depends on which developer you ask, but pretty much they are all in agreement the code base needs to be cleaned up.   Different issues are affecting different things.    The size of some of the files is causing major issues with scoping right now.   There is a very real desire by other developers to shift to a class based hierarchy.  The size of files and not having functionality grouped together is complicating that shift.

The fact we are not modulized and some of the files have grown to huge sizes also makes it difficult to stay out of each other's way when there are multiple areas being worked on simultaneously.

**Comment by gddeen** [comment]
That's kind of the part I'm missing, the way people are getting trounced upon. The github system handles the merges. There would be interactions when "they unfortunately picked the same spot" to add code, or changed the same content. Moving to objects "might" prevent both of these, but probabilistically only the first.

Objectizing "additional" features is a great idea but may preclude the lower power products if the objects "perform" adjustments to the base product. The code is structured as is to reduce cycles. Objects aren't inherently slower but the software industry is packing up shop to switch to a larger 64 bit product. People writing for this class of product do not normally show interest in performance, in my opinion. If you jump to use the big tools/ideas, the product will end up with the fully protected execution environment. Bang, 8 bits may be too slow. (I have no concern one way or the other.)

If I were a business, I would look to the future and see if Marlin fits into the rollout of those cheap, slow 4.3GHz SOC 8 core 64b processors while maintaining dominant presence in the current low cost, hobbyist world.

Is that future a smooth transition. Are the processors of the near term incrementally better or do they require different OS. (I see Marlin as an OS.) Then, the question is build or buy the OS, build or buy the functionality.

A brand spanking new preemptive, multi process, multi threaded "built" Marlin is AWESOMELY fun to create. (having worked on non-virtual realtime dispatcher/scheduler in the far past.) It brings in a class of people that are fun to work with.  But, it is highly selective. This is kin to writing an Amiga ('85) OS and getting all the platforms to cross compile to it. Not THAT tough. You can end up with a good real-time OS with reduced complexity but is integrated into the major development platforms. It may take months.

Or, you can "buy" (not the exchange of legal tender) one that fits perfectly or can be modified to a "Androidish" look and feel. Maybe some have been abandoned by their companies? (There are many to consider.) 

I (myself) wouldn't radically change the design without some real consideration of where you'll end up.
I think the change is coming quickly and the TTL (dual pun. TimeToLive/TransTransLogic) days are numbered. 

Unix built up into a big protected user environment. Maybe it can also build up into one that also supports a large real-time environment aimed at something slightly easier than the requirements of a unix kernel developer. I don't have difficulty at all seeing python, shell, java in the mix, but I wouldn't want it interfaced to real time hardware. So, maybe you're looking for a mix. A simple easy to use arduino like environment in the kernel and the user space.  (There are many many reasons to dig into everything I've typed. But I'm sure lots of people are interested.)

Soapbox off...

**Comment by Roxy-3DPrintBoard** [comment]
> That's kind of the part I'm missing, the way people are getting trounced upon. The github system handles the merges. There would be interactions when "they unfortunately picked the same spot" to add code, or changed the same content. Moving to objects "might" prevent both of these, but probabilistically only the first.

In theory, Yes.   But in practice it doesn't.   A good example would be the Auto Bed Leveling code.    Currently, we have  None, 3-Point, Mesh, and Grid.   Each of these has various flavors and targets.   When somebody changes something simple like what happens when the Z-Probe is retracted but still triggered, chaos breaks out.    You start to layer different machine types on top of the Auto Bed Leveling and you really have a complicated stew to pick through.   And then you add the fact nobody has each machine type and it is real easy to understand how problems creep in.

The code needs to be simplified and cleaned up.

>  If you jump to use the big tools/ideas, the product will end up with the fully protected execution environment. Bang, 8 bits may be too slow. (I have no concern one way or the other.)

> If I were a business, I would look to the future and see if Marlin fits into the rollout of those cheap, slow 4.3GHz SOC 8 core 64b processors while maintaining dominant presence in the current low cost, hobbyist world.

We are pretty careful to keep the code efficient.  I think the 8-Bit AVR's still have some life left in them.   But it really doesn't matter.  The high performance 64-Bit AVR's are starting to be available and will be very inexpensive soon.   There are already Marlin ports to the 64-Bit AVR's.  We probably will be able to support both with the same code base for some time.  It may be some of the fancier stuff like full graphics LCD Panels only run well on the 64-Bit AVR's.   I don't know what compromises we will have to make, but I'm pretty sure we can adjust.

We already have some use of C++ classes  in Marlin.   They haven't been too hard on the processor.  But we will have to be careful about that as we introduce more class hierarchies.

**Comment by gddeen** [comment]
Ah Ha! Short term consideration! Not strategy session.
The current arduino systems are plenty fast to handle all C++ objects. Just place them in non-realtime path as they are today.

- Appoint a Marlin god. (Probably is one.)
- Mesh/grid both use the same concept. Grid has 1 plane. Mesh has N planes.
  Make a “Plane object.” Grid will pretty much use one, mesh lots. 
  (Grid is simpler in implementation, but pretty much bent my x bar. It’s very easy to make it into object.)
  (Mesh actually works - Keeps bed from getting crashed.)
  Make 1 object everyone uses, but can select multiple squares/points, start/call mechanisms.
- Anoint a new object “Probe Object,” subclass into “Probe end stop object"
- Anoint a new "End stop Object”
- Anoint a new “Coordinate object”
- Annoint a new “Board pad mapper to logical address to hardware address mapper”
- Design a “printer feature” object. Select a feature on or off, dynamic or static.
- Design an “old horrible display” object.

Hammer out the object design. Pass them out.
Get everyone to agree on the objects and how they work.
Let people redesign their objects. Meet again.
Go grab steal the current code. Put it all together.
Require it all compile as C++.

I would assume using an Object to describe my printer is just as easy as configuration.h file. Manually no big deal for average user.

I would assume a grand object based Marlin everyone wants.

I would assume a grand poobah is needed to keep stuff working together.

I would assume a new “use this object” instead of that object would become the common desired practice.
(His object doesn’t work, hers does, etc…)

I would assume business as usual on bug fixes, enhancements, some features.

I can hardly wait!!!

**Comment by jbrazio** [comment]
The problem with objects is I believe it's incompatible with the current Marlin dev workflow. Objects would require everything to be designed and planned upfront implementation, this is not the current Marlin philosophy.

**Comment by Roxy-3DPrintBoard** [comment]
Wow!   

Perhaps a counter example will help the discussion:   Would a Z-Probe object require everything to redesigned prior to it's instantiation?   My guess is a Z-Probe object could be made that supported things such as Deployment(), Retraction(), Run_Z_Probe(), Report_Repeatability(), Set_Servo_Angles(), etc. without redesigning everything else. 

And if this Z-Probe object existed, it would be almost trivial to use it instead of the existing calls.   As @gddeen suggested, the Grid, Mesh and 3-Point Auto Bed Leveling have a lot of common usage of Planes in their functionality.   This is more tricky because the real time code needs fast and easy access to underlying data.  We would have to be especially careful how we constructed an Auto_Bed_Leveling object for this reason.   But I think we can see a few methods and class data that all flavors have.

**Comment by thinkyhead** [comment]
Stepper, planner, temperature, and other sub-modules can today quite easily be wrapped up into singleton classes to get the encapsulation we like, and without any performance hit, because we would not be using any fancy features of C++. This is a process that will unfold gradually. The functionality and contents of the functions of the initial classes would follow the same pattern as the current code. But, as time goes on and more things –like endstops, extruders, etc.– become encapsulated into objects, the structure will begin to fall into place more. I don't think we'll see any major overhaul of the execution flow. But we will have nicer encapsulation and the code will be getting easier to follow over time.

**Comment by gddeen** [comment]
jbrazio. There are currently "objects" in Marlin. It's simply c++. One does not need to think and plan to create/design objects. However, objects can inherit "stuff" that just magically changes their behavior.
Maybe that isn't the smartest thing to do on an 8bit computer. So, don't do that.

Classes, objects all sorts of programming concepts were added in 80s for good reasons.
Hacking (Making) will still survive.

**Comment by CONSULitAS** [comment]
@gddeen
Perhaps you should think and have a look at examples for objects on Arduino before you disencourage people here by nonsense. Objects are not slow on an 8 bit machine! Stupid programming is slow.

If you don't believe i recommend some reading for you:
http://cosa-arduino.blogspot.com.es
http://cosa-arduino.blogspot.com.es/2013/03/benchmarking-pin-classes.html (slow objects?)
https://github.com/mikaelpatel/Cosa
http://forum.arduino.cc/index.php/topic,150299.0.html

Did you read it all?

So perhaps start contributing good code and show working concepts instead of bashing an important approach for making Marlin more maintainable and even more flexible.

BTW: I read it...

**Comment by jbrazio** [comment]
I'm not saying we should move to objects, it's a pleasure to read @repetier code. I'm just saying the with the technical shift a mindset shift must also happen.

A quick illustration of what I mean: a key feature of good OOP is self containment, meaning [as an example] global vars are over, if you need to access a particular data struct then this must be published by the object's public interface and no form of "hacking" should be tolerated from the developers. There is no more coordinate changes by just editing the current position array and then informing the planner.. we must tell the planner our intent and then it's up to it to do it and then inform us of the result.

**Comment by Sniffle** [comment]
Damn wheres a like button when you need it...

If your gonna go OOP do it right or not at all, maintain scope and keep
things private between objects so that the objects are just little black
boxes that take inout and give output.  I kinda have a feeling this is the
turning point for marlin.  Its either gonna grow back into the monster
powerhouse firmware it once was or its gonna fall into obsolescence.  This
is basically the step that allows marlins features to move forward onto 32
and 64 bit chips.  I realize im a very small player in this but you guys
also know i only open my mouth when i have a good point to make.

**Comment by CONSULitAS** [comment]
@Sniffle It is new just in the right upper corner of any comment shown as + :smile: .

**Comment by Roxy-3DPrintBoard** [comment]
These kinds of discussion have a tendency to become religious...  But...

> A quick illustration of what I mean: a key feature of good OOP is self containment, meaning [as an example] global vars are over...

No...  we need to be very careful with the words we use.   For example,  If we took the simple case of a instantiating a Probe Object, I'm pretty sure this object would be a global variable of class Probe with what ever appropriate extensions were needed by that particular hardware setup.   But...

> If your gonna go OOP do it right or not at all, maintain scope and keep things private between objects so that the objects are just little black boxes that take inout and give output. 

This hypothetical Probe Object would not have its class data available for anything else to mess around with it.   This Probe Object would be a little black box that just does what its supposed to do.

**Comment by Sniffle** [comment]
<3 roxy i just saw something worth talking about and threw in my .02 like
usual :-D

**Comment by jbrazio** [comment]
What happen to @Roxy-3DPrintBoard comment ?

> I'm pretty sure this object would be a global variable of class Probe

Of course we still have "global" pointers to objects, but they are no longer global, they are inside the **Controller** class.. in a nutshell what we know as `Marlin_main` will be a `Marlin` *controller* class which will take care of instancing all the other objects.

**Comment by Roxy-3DPrintBoard** [comment]
> What happen to @Roxy-3DPrintBoard comment ?

I deleted it because I'm probably the only one that doesn't have a formal Computer Science training.  I figured I would end up getting picked on because there is probably some technicality where an object is not considered a variable or some such non-sense!   :)

**Comment by Sniffle** [comment]
I only partially completed mine so im only good for theory.  I do
understand thats why i dont talk much.

**Comment by ghumphr1** [comment]
@Roxy-3DPrintBoard - I am an engineer - and i think you should have say whatever.... just because you havent such and such piece of paper doesnt mean you have no valid input.. more so than some of the 'qualified' people most times

**Comment by Roxy-3DPrintBoard** [comment]
> and i think you should have say whatever.... 

Everything is good.   And if you saw the deleted comment, the first line of it was "These types of discussions tend to turn religious..."    That is because the object orientated topic is very important with very strong beliefs on all sides.   

What we are all after is a cleaner and more easily maintained code base.  And probably some C++ objects can help on that topic.   But for me it isn't a religious matter.   It is putting the objects in places where it makes sense and helps encapsulate data and functionality.

But because Marlin has a bunch of real time code and is currently powered by a slightly under powered AVR processor, we have to be careful how we do it.  We can move in that direction, but it has to be carefully thought out.

**Comment by jbrazio** [comment]
As long as we don't go into the `malloc()` road..

If you go object oriented.. you should go all the way; the problem is when people forget they are on a AVR arch and start doing "dynamic stuff".

**Comment by ghumphr1** [comment]
* powered by a slightly under powered AVR processor*
lol..

**Comment by Roxy-3DPrintBoard** [comment]
> As long as we don't go into the malloc() road..

Another religious topic when it comes to firmware!  malloc() and free() can be used safely but they are so controversial, it is just best to avoid them if possible.

**Comment by jbrazio** [comment]
lol it's Easter, a lot of religious topics !

The "dynamic issues" come into plate when people start delegation into runtime stuff that should, in a embedded environment, be defined at compile time, some examples are number of end stops, number of extruders..

If we know upfront we how many extruder objects are needed then there is no need to start dynamically increasing the array where their pointers are stored just because we are dealing with objects, preprocessing is [and should continue to be] our valuable friend.

**Comment by thinkyhead** [comment]
The initial wrapping of Marlin elements into classes will be done without any changes to the functions themselves or the code outside the classes. It will just be a convenient wrapper so that we, fore example, only expose a single "stepper" object instead of all the global variables defined in `stepper.cpp`. Callers will simply be preceding all their function calls with `stepper.` (not `stepper->`). Stepper and Planner will be the first two objects created. The purpose of the first round of object-izing will not be to improve the design patterns, only to get encapsulation and limit access to data. It is only a first step, and out of it some additional order will emerge. If done gradually, Marlin will naturally morph into a proper architecture. But we have to take it slowly starting from our current minor mess, otherwise we might as well just start with another firmware that's already got this done.

Roxy, no worries over "skills." I have zero formal training in computer-science, and am entirely self-taught. Everything you need to know can be learned right on the Internet. The key thing is to pursue your interests, learn what you need to know as you go along, and keep trying new languages and environments. After a while, all programming looks a lot the same!

**Comment by Roxy-3DPrintBoard** [comment]
> Roxy, no worries over "skills." I have zero formal training in computer-science, and am entirely self-taught. Everything you need to know can be learned right on the Internet. The key thing is to pursue your interests, learn what you need to know as you go along, and keep trying new languages and environments.

Thanks for the encouragement.  Mostly what you are seeing is 'scar tissue' from previous interactions.  But that is behind us now.   I like messing with code that is 'down on the metal'.   (M100 for instance)   (and my current play toy the Delta Mesh Leveling System)

But when it comes to higher level architectural direction, my biggest contribution is going to be getting consensus from the people that really know the issues.  I am very serious about rebuilding the Marlin Team and making it a force to be reckoned with.

**Comment by jbrazio** [comment]
> The initial wrapping of Marlin elements into classes will be done without any changes to the functions themselves or the code outside the classes.

Absolutely agreed. I've been talking about high level design, not about the detailed roadmap.

`<offtopic>`
> After a while, all programming looks a lot the same!

Not all of [them](https://en.wikipedia.org/wiki/Brainfuck)..
```bf
++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.
```
(Impressive, Github has a bf syntax highlighter)
`</offtopic>`

**Comment by Roxy-3DPrintBoard** [comment]
~~~~
<more offtopic>
After a while, all programming looks a lot the same!
Not all of them..
++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.
(Impressive, Github has a bf syntax highlighter)
</more offtopic>
~~~~
So of course I get curious and start checking things out.  It says:

>A second version of the compiler used only 240 bytes.[4]

>As Aminet grew, the compiler became popular among the Amiga community, and in time it was implemented for other platforms. Several brainfuck compilers have been made smaller than 200 bytes, and one is only 100 bytes.

I was very disappointed.  It was nice that the source code for the compiler is available to look at.  But it is written in 68000 assembly language and not in BrainFuck.   It would seem to me the BrainFuck compiler should be written in BrainFuck.   And of course, this compiler had absolutely no error or syntax checking.

**Comment by thinkyhead** [comment]
I spent a few years writing 680x0 assembler on the Amiga, and I really liked that platform. I wrote a couple of games. I was the author of **Dino Wars** and the Amiga port of **Bill 'n' Ted's Excellent Adventure**.

**Comment by gddeen** [comment]
It WAS awesomeness.

**Comment by gddeen** [comment]
I am not a religious object fan anyway.  I am for what works the best. Nothing really works the best 
except the way the code is now.  The main feature are no big dynamic data. ConsultitAS thanks for
the reading material. I do think 16MHz machines are slow when compared to 4.4GHz 32 core ones.
Programmers these days think a quad core 1.2GHz 64 bit machine is slow. It's because they have
no control over all the excess luggage they carry.  I am a firm believer in "Keep it stupid, simple."
Elegance is nice, simple is nice but it needs to be understandable.

**Comment by Sniffle** [comment]
this developed into a bigger post than I intended, Ive been thinking on this a LOT lately...

/me puts foot on soapbox

So, IF marlin goes the object route or even if it doesnt, has anyone thought about laying out the current code, and deciding how it should be broken up?  The macro view of breaking the code up and planning for things to be converted, because without it, it's gonna be a dirty ugly mess as the current marlin dev code shows.  Before this happens unlike previous times there needs to be a clear plan of attack.  I know @thinkyhead probably has an idea in his head how he wants to go about this, and @Roxy-3DPrintBoard seems to be the driving force that is keeping this on track and has pushed marlin to the point it currently is and it seems she has more pulled people to the cause.  

so just for examples sake, things like how the code interacts, I'll give a code example and reasoning.  

I'm using this to learn C++ syntax and brush up on my coding skills so dont hold it to too high of a standard
'''c++
class probe{
	bool stowed;
	fake_actioner a;
	public:
		void extend(){
			a.extend();
			stowed = false;
		}
		void retract(){
			a.retract();
			stowed = true;
		}
		bool probe_state(){
			return stowed;
		}
		void probe_init(fake_actioner b){
			a = b;//needs to reference the original object not a copy
			this->retract();
		}
};
'''
Should the probe object just store its state or should it call the actioner to act upon it, or even further should there even be a probe object that stores its state, it technically is an object to be identified and fleshed out.

/me steps on soapbox
To me this is the elephant in the room, beyond breaking marlin up into modules.  You can want to break it up all you want to but without a cohesive plan of attack it's going to fall apart and I personally dont want that.  Deciding bug fix time is easy your not basically creating something new, your just fixing whats already there.  If marlin undergoes a large change like rewriting it as OO you might as well call it marlin 2.0 not 1.2 it is a complete rewrite of the old code to a new higher standard to increase code readability and functionality moving forward to be able to go to new hardware beyond the mega2560.

This is something that NEEDS to be be highly visible and enforced with big checkoffs and milestones to keep things on track no matter who comes and goes so that it will move forward no matter what, much like the coding standard that has been implemented.  I really like marlin and this community I want it to be so much more than it is...
/sigh
/me steps off soapbox

**Comment by Roxy-3DPrintBoard** [comment]
@Sniffle You have several questions.   First the easy ones:

> The macro view of breaking the code up and planning for things to be converted, because without it, it's gonna be a dirty ugly mess as the current marlin dev code shows. 

We already have several proto types for breaking up the files into modules.  In MarlinDev there is a good start at that.   We probably want to have a separate folder for the files that belong in each topic area.   Please take a look at the Feb14 branch.    

But...  I'm guessing this all gets abandoned.   MarlinKimbra has Due support running and has started breaking up the files into modules similarly to what you see in the Feb14 branch.   This next idea is going to be discussed to get everybody's input prior to it happening, but a number of people are in favor of taking the changes (improvements) in the Release Cadidate (Once it goes Golden) and crossing those over to the MarlinKimbra branch.   If we did that, we would use that as the starting platform for continued MarlinDev work.

> Should the probe object just store its state or should it call the actioner to act upon it, or even further should there even be a probe object that stores its state, it technically is an object to be identified and fleshed out.

The whole purpose of moving **SOME** of the functionality to being Object Orientated is to start isolating behavior and encapsulating data.   So, "Yes!", the probe object should store its state.  And this brings up an important consideration for the conversion process.   Each objects functionality (and data) need to be very quick and efficient to access.    We will have discussions on this, but we need things like Z_Probe->is_deployed().    But we may not want to incur a function call to get that information.   We may want to define an inline macro expansion of a is_deployed() function that goes straight to the class data that knows about the probe.   

The Z-Probe is a good example.   Its actual use is fairly limited.   And with the exception of the stepper code that is constantly checking to see if an end stop has been hit, its use is not in any critical real time code.   So, with the inline function work around for Z_Probe->is_deployed() I expect a Z-Probe object will be fairly clean and easy to transplant into the code base.    We cut out all of the old code, and just start referencing the base functionality of the Z-Probe object.

And at that point...   We have a single global pointer to a Z-Probe object.  And it can be defined to be any of the known types:   None, Manually Deployed, Manually Positioned Nozzle, Allen Key, Servo, Inductive, etc.  All of the quirks of particular probe will be out of the main code base and hidden in that particular probe's code.

> To me this is the elephant in the room, beyond breaking marlin up into modules. You can want to break it up all you want to but without a cohesive plan of attack it's going to fall apart and I personally dont want that. 

Well...   There is a lot of stuff that is obviously in need of clean up.   If we start getting the obvious things cleaned up, we will have time to think about and decide how to clean up the rest.   Right now, we do have a high level plan:   Start getting things cleaned up.       Later, we will have more specific thoughts on how to continue.

**Comment by Sniffle** [comment]
OK so I guess since I have a partially complete formal CS degree, I was taught that objects should be as simple as possible and should build and call upon each other to make more complex actions easier to follow and read.

I see a probe as an object that is either deployed or stowed, and it is acted upon by probe actioners that that define how that probe object is used.  Said actioners are the methodologies of deploying the probe, either by doing nothing or null object, manual deployment object, Manually Positioned Nozzle object , Allen Key object, Servo object, Inductive object, etc. and by having the probe do the calling of that actioner object, you have a probe object that can be acted upon by anything and will always store its state.  Below the probe object you would have a bed leveling object that is initialized by a general functionality object that inits a movement objects that init stepper objects... ...that init the HAL object that talks to defined hardware.  the config would be read by the general functionality object that goes on to init everything needed.  Partially going OO means that instead of referencing other objects to do tasks you are picking a spot to remove all the layers below and dump them all into that object much like how marlin_main.cpp has gotten so large, at that point in the code things suddenly get complex, bloated and hard to read with code duplication everywhere..  

Using this type of methodology for converting the code if there was ever an issue with say the allen key code you would be able to go to the allen key actioner directly and only deal with that clearly defined piece of code never actually needing to touch the probe code or anything else.  All of those objects could be either broken out into their own files or be a single probe actioner file with clearly defined and hidden methods.

I just have a hard time seeing how you can go partially OO, Its so integral to the way things are coded and how the code flows partiality will probably cause a lot of issues.  I guess where I am getting hung up is the actual coding methodology for implementation vs the concept and plan for doing so.  I'm really not trying to be a naysayer, I want it to happen and I'll help as much as I possibly can, but wrapping my head around how to do it without starting over from the bottom and working our way up without a solidly defined plan is giving me headaches.  As an idea, go look at Smoothies source and how it is broken up using full OO standards(as far as I can tell.) It is a good place to look for examples on how OO works out and why partiality is hard to do. wolfmanjm is a hard taskmaster for keeping smoothies code on track, elegant and properly developed.


again these are all observations on the assumption that the marlin developers will choose go OO

**Comment by bobc** [comment]
Breaking the code into modules is a great idea, it's an essential addition to the Arduino IDE IMO.

As regards OO, I did a CS degree but also learnt through job experience that KISS and being pragmatic are more important in reality than maintaining a pure OO abstraction. I've seem some godawful code written according to a supposed "OO paradigm".

I prefer evolution rather than the big bang approach. As (I think) all of the codebase is actually compiled as C++, there is no linking issues introducing objects in stages.

The first step to producing a clean OO abstraction is to figure out what the modules are, and what the dependencies are. Some refactoring will certainly be needed.

**Comment by jbrazio** [comment]
@Roxy-3DPrintBoard @thinkyhead seems that we're the only ones without a formal CS degree around here. ;-D

**Comment by Sniffle** [comment]
I never finished mine :-D so you arent alone :-P

**Comment by bobc** [comment]
Heh, my degree predates most of the OO stuff any way. Not much call for FORTRAN any more...

**Comment by gddeen** [comment]
There is PLENTY of call for FORTRAN still (like it or not...) You could probably translate libraries to APL?(jk)
Formal degrees are only good when the person dedicates themselves to them.
In programming, daily usage is catamount. Unless you want to know if a problem is NP complete, left or right recursing grammar, or why the heck random numbers are so necessary to computational science.

**Comment by CONSULitAS** [comment]
@jbrazio @Roxy-3DPrintBoard @thinkyhead 
OK: I drop my pants, too. I only have an PH D in chemistry.

But: Does some C source and Bash scripts in the appendix of the thesis (found online) count? Perhaps not. 😁 😀 

But: The formal CS degree says nothing about a programmer or IT architect. I wrote my first program down in 1980 and had my first own computer in 1981. Have there been CS degrees in germany at all? Dunno exactly. I have been much to young for such stuff 😄 .

@all
So lets get back to the topic. IMO:
* We need modules, because Marlin is not as structured as it should be.
* We need modules, because it makes Marlin more maintainable.
* Objects are great and should be used. But step by step and function by function.
* If you read about the object implementation in C++ it is just an enhanced `struct`. So a good step to start is to pack variables which belong together in structs.
* For this struct you can then start to write functions working with it aka "methods".
* If an object is not completely ready, you could make the properties you cannot access by methods public. So legacy code sees just a struct without knowing about the object.
* Big bang never works when refactoring. So find a way and start walking. Step by step.

So lets start!

A really annoying part of code which makes a lot of trouble for users is EEPROM.

Take a look at https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L148 and following.
* Imagine, how it would look like, if we had a struct with all feedrates, all accelerations and all jerks. Just a few lines left.
* A major advantage: If there is a new `max_f_jerk` e.g. for a new tool, it would work without a change!

And now https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L504 and following.
* If the struct would be extended to an object that has a method named `resetToDefaults`, you would just write `feedrates.resetToDefaults();`.

Interesting is the part https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L515 
* I would suggest `bedLevelingMesh.disable();` and if the feature is switched of, all methods are exchanged by dummies. So you wouldn't need an `#if`, because it sounds like:

```
bedLevelingMesh.disable() {
  #if ENABLED(MESH_BED_LEVELING)
    bedLevelingMesh.active = 0;
  #elsif
    // do nothing, because no action required if MBL disabled
  #endif
}
```

BTW:
* The call of the now empty method `disable()` will be optimized by the compiler. So the resulting code is the same, but shorter, better to maintain, less nested `#if`s and so on.
* I would recommend to define a naming convention for any object or struct in Marlin: Important part first and camel case e.g.
  * bedLeveling
  * bedLevelingMesh
  * bedLevelingManual

So you can better sort, find, see parts that belong together. And it fits the way C(++) uses for structs.

So https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L533 would result in a struct like
```
// anywhere
#define PLA 0
#define ABS 1

// resulting code
preheat.temp.hotend[PLA] = PLA_PREHEAT_HOTEND_TEMP;
preheat.temp.heatbed[PLA] = PLA_PREHEAT_HPB_TEMP;
preheat.speedFan[PLA]
preheat.temp.hotend[ABS]
preheat.temp.heatbed[ABS]
preheat.speedFan[ABS]
```

Than https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L206 would be funny. Just:
`EEPROM_WRITE_VAR(i, preheat);` or perhaps `preheat.eepromWrite();` which would use an EEPROM object and just inherit the code from a generic configuration object. And then every new object has to have its own logic for writing GCODE in `Config_PrintSettings`.

As an result i would extend `configuration.eepromWrite();` to `configuration.eepromWrite(mode);` with modes for EEPROM, GCODE, SD and more to come. One code fits all. And you can keep it quite simple.

Not that challenging, but a lot of work and changing the look and feel of Marlin quite a bit.

Uh! Quite a long contribution. What do you think?
* Step by step starting with structs and easy objects?
* Start with configuration variables as global variables which should be encapsulated and seem to be rather simple?

**Comment by MagoKimbra** [comment]
Look this https://github.com/MagoKimbra/MarlinKimbra4due/tree/V4_3_0_new_tools/MK4due/module/tools

Test version width object Hotends and Beds...

**Comment by bobc** [comment]
@MagoKimbra Ok, seriously now, what the fuck are you doing?

You've taken other people's code, removed their names and put your name on it. This is just not fucking cool.

Unless there is *very prompt* action I shall be filing copyright complaints.

**Comment by jbrazio** [comment]
@CONSULitAS thanks for your wise opinion.

It's almost nothing, but I've converted the SD/Serial print timer into an [object](https://github.com/MarlinFirmware/Marlin/blob/RCBugFix/Marlin/stopwatch.h), I was trying to eyeball what would be the acceptance of such approach: start migration small parts of the code into dedicated objects. @thinkyhead approved the approach but I didn't received any other feedback, now I see you do support the same approach for the refactoring. :+1: 

> Imagine, how it would look like, if we had a struct with all feedrates, all accelerations and all jerks. Just a few lines left.

As entry points to the `ConfigurationStore` read and write are quite easy to pin point on Marlin, IMO this one is a perfect [next] target to be "objectified".

Now imagine we, on the new `ConfigurationStore`, implement a `ConfigStore.export();` to a SDCard and the all the SDCard stuff is still as is today.. this is my approach issue: when we have "objects" that need to access parts of Marlin which we have no refactored yet. Should we leave them as they are and later on revisit them or should we start chaining them together ?

> If the struct would be extended to an object that has a method named resetToDefaults, you would just write feedrates.resetToDefaults();.

Are you suggesting to have multiple structs, i.e. one for feedrates, another for accelaration, jerk etc ?

> and if the feature is switched of, all methods are exchanged by dummies

In order to be able to do `mbl.disable()` it means that the `mbl` object would always need to be defined, independent of `MESH_BED_LEVELING`, wouldn't this cause impact on the memory footprint if we transpose to "all" the other Marlin features ?

> I would recommend to define a naming convention for any object or struct in Marlin

Not crafted in stone but I've been told to use `UpperCamelCase` for objects and `lowercase` for file names, nevertheless I would agree `lowerCamelCase` looks better for methods.

> And then every new object has to have its own logic for writing GCODE in Config_PrintSettings.

Are you suggesting each object which has EEPROM data would require someting like `printSettings()` method to dump it's own settings and then `ConfigurationStore` would do something like this ?

```
mbl.printSettings();
preheat.printSettings();
(..)
```

**Comment by CONSULitAS** [comment]
@jbrazio
> thanks for your wise opinion.

You are welcome. BTW: Thanks for all the great work you did on Marlin in the last weeks.

>It's almost nothing, but I've converted the SD/Serial print timer into an object

No, this isn't nothing, its just one step into the right direction. It will make code much more capsulated and readable. I had a short look at it and think it is rather nice now! But you should perhaps change `status` to `state`, because you implemented it as an state machine, which is desirable, because they are rather stable and easy to understand. 

>  @thinkyhead approved the approach but I didn't received any other feedback

* Have a look which members of the core team are active (and merging) in the last weeks. You've  got a good statistics. 😄
* You could say in this moment: "It is good code, if @thinkyhead merges it."
* And with every community work: If nobody complains loud, it is in most cases a great solution!
* **Perhaps just start a PR with your little nifty stopwatch and see whats happening.

> As entry points to the ConfigurationStore read and write are quite easy to pin point on Marlin, IMO this one is a perfect [next] target to be "objectified".

Thats exactly the reason why i picked this part of code. For the start of refactoring you should use simple examples, because we have to learn and exercise. So it has to be easy.

> Should we leave them as they are and later on revisit them or should we start chaining them together ?

If you try to solve every dependency, you will find, that you have to change everything in a big bang. So **leave them as they are**. This is important for bug fixing, too. If you change to much, you don't know where to search.

> Are you suggesting to have multiple structs, i.e. one for feedrates, another for accelaration, jerk etc ?

Not really. It has been a quick and dirty example just before going to bed. But thinking more about this topic, it makes sense to chain every config to the object it is needed for.

So we have to think which objects make sense. I would suppose to start with the simple ones again to make experiences.

> wouldn't this cause impact on the memory footprint if we transpose to "all" the other Marlin features ?

If we define it only with empty dummies and without properties it will be optimized away and don't hurt.

> Not crafted in stone but I've been told to use UpperCamelCase for objects and lowercase for file names, nevertheless I would agree lowerCamelCase looks better for methods.

I would agree, but for different reason:
* Objects are things with names and you write names starting with capitals like `João`. So my object should be called `Preheat` instead.
* Methods are instructions what to do and you can count them as verb. This is written lower case like `goOn`.
* `João.goOn();` is readable like a sentence, what is probably the reason why you think it "looks" better.

> Are you suggesting each object which has EEPROM data would require someting like printSettings() method to dump it's own settings and then ConfigurationStore would do something like this ?

Excactly. And the objects than should use a public method of the EEPROM object, because this is the master of the storage and encapsulates the technical interface.

**As an result you get the long searched for HAL!** If the EEPROM object knows how to interface with the hardware and the configuration items just use it, you can exchange the code as you like, if you only keep the interface.

So the EEPROM object should be renamed to something like PersistentStorage which supports channels like SD, EEPROM, GCODE, engraving to stone plates 😁 or anything else. It has just to be persistent (in a way).

But this leads to the point, where you need to redefine the way you store the information. If you want to write GCODE and binary data. Look here;
https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L614

I would prose something like:
```
Axis.printSettings();

in Axis.cpp:

Axis::printSettings() {
...
PersistentStorage.saveData("M92 X",axis_steps_per_unit[X_AXIS]);
PersistentStorage.saveData("M92 Y",axis_steps_per_unit[Y_AXIS]);
PersistentStorage.saveData("M92 Z",axis_steps_per_unit[Z_AXIS]);
PersistentStorage.saveData("M92 E",axis_steps_per_unit[E_AXIS]);
...
}
```

This looks similar to the actual code, but the result is completely different: M5xx only calls the objects to be saved and the objects care for their selves. And if anybody extends the object like adding a new F axis he will see the method and adapt it. Now nearly every time EEPROM code is forgotten.

Sorry, but i didn't have the time to make it shorter. 😀

**Comment by Roxy-3DPrintBoard** [comment]
> Are you suggesting each object which has EEPROM data would require someting like printSettings() method to dump it's own settings and then ConfigurationStore would do something like this ?

I'm wrestling with a similar question for the High Resolution Delta Mesh Bed Leveling.  I'm going to be storing a matrix of numbers in the EEPROM and reloading at reboot time.   What I was going to do in order to not be very invasive was have a method inside my Auto_Bed_Leveling object that would write its data at an index passed into it.  And it would return the index of where subsequent EEPROM writes can occur.   

That would allow me to just add one line in the Configuration_Store.cpp file to save my object's data, and another similar method to restore its data.

**Comment by jbrazio** [comment]
I'm on mobile right now and I need a PC to fully answer to the previous two comments but how about a EEPROM_lookup_t[] which hold data structs, each struct would define:
 - number of bytes reserved
 - function to be called for get/put data into the object

Like this ConfigurationStore would be agnostic and would not require to be edited each time a new EEPROM data block is introduced, just add it to the table.

ConfugurationStore when reading the data from EEPROM would check the offset and call the required object function which will then do whatever is required with the data.

IMO this goes into the direction @CONSULitAS was speaking about "isolating" each object and their data into a container, making each object the sole responsible for its footprint.

**Comment by Roxy-3DPrintBoard** [comment]
> but how about a EEPROM_lookup_t[] which hold data structs, each struct would define:
* number of bytes reserved
* function to be called for get/put data into the object

Or...  Maybe we can free ourselves from fixed locations in the EEPROM.  Maybe each object that wants to store data can give the EEPROM (or Persistent_Storage object) the name of the data that it is storing.  Then as items get added or deleted, it just doesn't matter.  The previous values can still be found and used.

**Comment by jbrazio** [comment]
At first look I see a possible issue because the order to which each object registers itself with the `ConfigurationStore` would then matter, or am I under seeing this ? But this type of "freedom" would be great I do agree with you.

**Comment by Roxy-3DPrintBoard** [comment]
Well...  This might not make sense to do...  But if we had each object say to save its information with a name, the Persistent_Storage object would keep a directory where everything was.  It would save a list (that nobody else sees) telling it the location of where things had been saved.  Data would float depending on what got saved and even the order things got saved.

**Comment by CONSULitAS** [comment]
@Roxy-3DPrintBoard 
> write its data at an index passed into it

Please don't do so. You would loose HAL! The "BedlevelingDeltaMesh" object may not know about how and where to store data. It has just to pass all data to an universal interface.

> That would allow me to just add one line in the Configuration_Store.cpp file to save my object's data, and another similar method to restore its data.

Thats a main point of the idea. BUT: Only one method to read, store, reset and so on for the data. Look at the three ones - most of them is completely the same.

@jbrazio 
> but how about a EEPROM_lookup_t[] which hold data structs, each struct would define:

You would have to configure it. Make it self configurable.

> Maybe we can free ourselves from fixed locations in the EEPROM. 

Thats what i am speaking of. With fixed locations you would loose HAL, too! There are two possibilities:
* Think of it as reading an GCODE. Parsing the data for a key and reading the data to the correct object. Ever heard of fixed locations for GCODE? 😀 
* Or write GCODE to every location
  * SD `BootConfig.g`- Advantage it is human readable AND editable.
  * EEPROM - check how much bytes the GCODE written by Config_PrintSettings() https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L605 is. There is plenty space.
  * Serial - We need it for M503
  * perhaps I2C EEPROM for Due?

> At first look I see a possible issue because the order to which each object registers itself with the ConfigurationStore would then matter, or am I under seeing this ? 

No, thats the point. This has IMO one central advantage: You can have new features, you can read partial configs (just init to default like M502 and then read for M503) and the new features are at least configured to default settings. All because the data is not only "binary garbage" if you shift the order of things to be stored.

Let me sketch the algorithm:
```
enum DataMode {ModeWrite, ModeRead, ModeReset};
enum DataTarget {TargetNone, TargetSD, TargetEEPROM, TargetSerial};

ConfigurationStore::writeData(target) { this->handleData(ModeWrite, target); } // aka M500
ConfigurationStore::readData(target) { this->handleData(ModeRead, target); } // aka M501
ConfigurationStore::resetData() { this->handleData(resetMode, null); } // aka M502
ConfigurationStore::M503() { this->writeData(TargetSerial); }

ConfigurationStore::handleData(DataMode myMode, DataTarget myTarget) {
  switch (myMode):
    case ModeWrite:
      PersistentStorage.open(myTarget); // Open File / write or read header / init checksums / do magic stuff
      BedlevelingDeltaMesh.writeConfig(myTarget); // let the object select it's data
      Axis.writeConfig(myTarget);
      ...
      PersistentStorage.close(myTarget);
      break;
    case ModeRead:
      // perhaps just reading data and passing to the GCODE parser (after checks)
      ...
    case ModeReset:
      BedlevelingDeltaMesh.resetConfig();
      Axis.resetConfig();
      break;
    default:
       // should not happen!
  }

Axis::writeConfig(DataTarget myTarget) {
  // compare https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L614
  PersistentStorage.write(myTarget, "M92 X");
  PersistentStorage.write(myTarget, axis_steps_per_unit[X_AXIS]);
  PersistentStorage.write(myTarget, " Y");
  PersistentStorage.write(myTarget, axis_steps_per_unit[Y_AXIS]);
  PersistentStorage.write(myTarget, " Z");
  PersistentStorage.write(myTarget, axis_steps_per_unit[Z_AXIS]);
  PersistentStorage.write(myTarget, " E");
  PersistentStorage.write(myTarget, axis_steps_per_unit[E_AXIS]);
}

Axis::resetConfig() {
  // just copy https://github.com/MarlinFirmware/MarlinDev/blob/dev/configuration_store.cpp#L488
}

inline void gcode_M92() { Axis.readConfig(current_command_args); } // still worth an own file, module or even an function? Yes, but only for structure reasons. :-)

Axis::readConfig(char command_args) { // handling M92
  // just copy https://github.com/MarlinFirmware/MarlinDev/blob/dev/Marlin_main.cpp#L4248
}
```

BTW: This are all not brand-new ideas for me. I've been thinking about our EEPROM problems at least since this post: https://github.com/MarlinFirmware/Marlin/issues/1398#issuecomment-71169003

Any ideas?

**Comment by CONSULitAS** [comment]
one more point: 😄 
```
PersistentStorage.write(DataTarget myTarget, char myCharToWrite) {
  switch (myTarget) { // sic!
    case TargetNone:
      break; // do nothing, dummy target
    case TargetSD:
    case TargetEEPROM:
      EEPROM.writeByte(myCharToWrite); // EEPROM object has to care for position, checksum etc.!
      break;
    case TargetSerial:
      SERIAL_ECHO(myCharToWrite);
      break;
    default:
      // oops!
}

PersistentStorage.write(DataTarget myTarget, char *myStringToWrite) {
  int size = strlen(myStringToWrite);
  for (int pos=0; pos<size; pos++) {
    this->write(myTarget, (uint_8t) myStringToWrite[pos]);
  }
}

// and so on for the important data types
```

**This is the HAL step!** So all high level methods don't care for the target, just the low end ones with the handling of single bytes. And this is really easy on Arduino. 👍

**Comment by Roxy-3DPrintBoard** [comment]
I'm thinking along those lines...  But it would be good for each object to be able to easily load or save its data.   If each object kept its savable data in a structure, then each object could easily save and restore its data:
~~~~
ConfigurationStore::writeData(target, char *object_name, char *data_ptr, int length_of_data_struct);    // aka M500
ConfigurationStore::readData(target, char *object_name, char *data_ptr, int length_of_data_struct));   // aka M501
~~~~

**Comment by CONSULitAS** [comment]
@Roxy-3DPrintBoard 
But why should an object load or save its data?
* On boot up? We call each object.init(); and it calls its own resetConfig() besides other todos.
* On GCODE? Which are the ones we have? The M5xx we have defined. Other GCODES should not have any surprises by side effects which change configuration permanently! 
* On shut down? People just switch off their printers.

**Comment by Roxy-3DPrintBoard** [comment]
> But why should an object load or save its data?

Let's take an extreme case of a High Resolution Mesh Bed Leveling System.   I'm going to allow a 15x15 grid that is easily edited but it will still take time to get it setup and fully calibrated.  The user is not going to want to generate the mesh every time they print.   So...  This Auto_Bed_Leveling object is going to need to be able to save and restore its data.

And...   There will be multiple ways to edit the mesh data.  The easiest will be from the LCD Panel.  But I think it also makes sense to let it be edited with sub-codes on the G29 command.

Maybe I'm just not understanding your question and concern.   Certainly, a M500 should result in each object that has data doing what ever we decide is appropriate to get its information saved.   And once it is saved, when the system boot's up...  The constructor for each object should be smart enough to go out and retrieve its data from the persistent storage facility.

**Comment by jbrazio** [comment]
@CONSULitAS, @Roxy-3DPrintBoard 

> engraving to stone plates

Punch cards, do not forget punch cards ! :tada: 

> PersistentStorage.saveData("M92 X",axis_steps_per_unit[X_AXIS]);

Why co-relate the G-Code with the data into the `saveData()` method ?
And how exactly would that translate into the persistent memory layout ?

> // perhaps just reading data and passing to the GCODE parser (after checks)

and

> But why should an object load or save its data?


Well if we want to let each object to be responsabilize by it's own data then we have to deliver the data back to the object, I would prefer this type of "container", otherwise we have to write validation code inside the `ConfigurationStore` or maybe even worse things.

In a nutshell: Each object must be responsible for it's own data as **only him knows** how to handle the data, but to save/load it own data the object must have a little help from it's good old pal `PersistentStorage`, as **only `PersistentStorage` knows** how to physically write/read the data [and so on].

**Comment by Roxy-3DPrintBoard** [comment]
As a related thought about objects I started wondering about some issues.   My guess is most of the code we convert over to objects would end up as global objects.   If this is true, I have some questions:

If our objects are global, how do they get instantiated?  Does that automatically happen by the C run time libraries?  I don't think it can happen at this time because Marlin would not even be initialized yet.

Maybe the global objects get constructed the first time they are used???  But even that possibility leaves me scratching my head, because the Persistent_Storage object would have to be alive and well for it to help out the other objects retrieve their data.   So, do we have to control the order of the instantiation of the objects?   Or do the constructors just start calling everybody further down the chain?

**Comment by CONSULitAS** [comment]
@jbrazio 
> Punch cards, do not forget punch cards ! :tada:

:grinning: Not kidding: Exactly thats the point. :+1:  It has to work with punch cards to be hardware independent! If you abstract the hardware to the max it works with stone plates and punch cards. If it doesn't, it's only partly HAL:

But to describe it a bit more seriosly:  
> And how exactly would that translate into the persistent memory layout ?

It doesn't! Because if you abstract a permanent storage totaly from hardware, you get an abstract data stream without layout - just serial. This is the smallest common sense of storage.

Sure, i would write a header with magic token (like `#!GCODE\n`) and a tail with number of bytes, checksum or something like this to eeprom. But i wouldn't call this "memory layout". Perhaps we should add the token to GCODE on other technologies as well? We should. But the checksums are surely technology dependent (not for SD card).

BTW: Do i now need to give attribution by beeing inspired by `#!/bin/bash`? :grimacing: :grinning: :cry: 

And this solves one of our serios problems: New versions for every new parameter, loosing all config data if you upgrade eeprom version (and the normal user will not recognize and understand), need for and maintaining an layout, garbage if you compile with different config ( :-1: sic!) and so on. 

If you don't want to invent the wheel again, you find a serial output of config data: M503

So the idea is:
* We skip every layout, because it doesn't work and makes users angry.
* We store GCODE in ASCII on every technology.
  * perhaps with checksums
  * on Storage with plenty space (SD, Serial, ...) perhaps with comments (if easy to implement)
  * use existing (and working!) code - just wrap it nifty!
    * GCODE parser for reading
    * old M503 broken in pieces for writing
  * make it human readable
    * how about a `M504` just dumping PersistentStorage to Serial for debugging?
    * edit your own Config manually on SD 
    * echo the GCODE to serial while doing M500/M501 -> great for debugging
  * Every object that want's to store data needs a GCODE for setting this data - like `M92`.
    * If it doesn't it will need one - or you could'nt change config by terminal
* Each object cares for its own config, but coordination with
* ConfigurationStore and
* PersitentStorage as HAL/driver

KEEP IT SIMPLE, STUPID! And i am sure, if done right it will save PROGMEM.

> Why co-relate the G-Code with the data into the saveData() method ?

Just, because we will store GCODE in ASCII in every persistent storage. It works, its there, its needed. Why invent a second format for the config on other technologies?

There are users that don't use the EEPROM and just put M503 output to their SD card. We would automate this and generalize it for any technology.

> Well if we want to let each object to be responsabilize by it's own data then we have to deliver the data back to the object, I would prefer this type of "container", otherwise we have to write validation code inside the ConfigurationStore or maybe even worse things.

No, please look at this code:
> ```
> inline void gcode_M92() { Axis.readConfig(current_command_args); }
> // still worth an own file, module or even an function? Yes, but only for structure reasons. :-)
>
>Axis::readConfig(char command_args) { // handling M92
>  // just copy https://github.com/MarlinFirmware/MarlinDev/blob/dev/Marlin_main.cpp#L4248
>}
>```

We just give the object the parameters of its read config GCODE.

> In a nutshell: Each object must be responsible for it's own data as only him knows how to handle the data, but to save/load it own data the object must have a little help from it's good old pal PersistentStorage, as only PersistentStorage knows how to physically write/read the data [and so on].

Correct, thats the idea: But we have layers of objects, and you forgot ConfigurationStore:
* `ConfigurationStore` as the manager / coordinator
  * calls `Axis` and all other configuration items -> mostly physical objects / worker
    * calling PersistentStorage as the servant / coworker

ConfigurationStore
* `PersistentStorage.init();`
* cares for the events and not to forget one -> n calls
* `PersitentStorage.close();`

Axis - and others
* cares for data and its interpretation

PersistentStorage
* cares for streams (files, serial, ...), integrity (errors, checksums, ...), headers, tails, hardware interface

@Roxy-3DPrintBoard 
> If our objects are global, how do they get instantiated?

You are right. Physical objects, that exist only once in reality are commonly instatiated as singletons. Becaus you need it once (only one EEPOM), its all the time there. Look at driver implementations. E.g. Watchdog in Cosa for Arduino. They will be initiated on boot before main().

**Comment by bobc** [comment]
> Just, because we will store GCODE in ASCII in every persistent storage. It works, its there, its needed. Why invent a second format for the config on other technologies?

SD card : 2,000,000 KB
EEPROM: 4 KB

Have you got a miraculous new compression algorithm to share ? :)

> Or do the constructors just start calling everybody further down the chain?

Constructors are not called in any particular order according to the C++ standard, but there is a GCC extension to assign priority levels to constructors.

**Comment by CONSULitAS** [comment]
@bobc 

You are kidding! Did you ever have a look at the output of `M503`?

My current config printed as `M503` Serial Output:
```
	echo:Steps per unit:
	echo:  M92 X128.50 Y128.50 Z2560.00 E1370.00
	echo:Maximum feedrates (mm/s):
	echo:  M203 X500.00 Y500.00 Z5.00 E25.00
	echo:Maximum Acceleration (mm/s2):
	echo:  M201 X9000 Y9000 Z100 E10000
	echo:Accelerations: P=printing, R=retract and T=travel
	echo:  M204 P1000.00 R1000.00 T1000.00
	echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
	echo:  M205 S0.00 T0.00 B20000 X20.00 Z0.40 E5.00
	echo:Home offset (mm):
	echo:  M206 X0.00 Y0.00 Z0.00
	echo:Material heatup parameters:
	echo:  M145 S0 H190 B50 F0
	echo:  M145 S1 H240 B60 F0
	echo:PID settings:
	echo:  M301 P24.29 I1.58 D93.51 C100.00 L20
	echo:  M304 P341.88 I25.32 D1153.89
	echo:Filament settings: Disabled
	echo:  M200 D3.00
	echo:  M200 D0
```

```
wc -c MarlinConfig_2016-04-16.txt 
     842 MarlinConfig_2016-04-16.txt
wc -c MarlinConfig_2016-04-16_gcode
     324 MarlinConfig_2016-04-16_gcode
```

Included `echo:`, Comments, Newlines we are below 1k. Without them (pure GCODE) we are below 400 Byes. We have 4k. We could even skip spaces to save some bytes...

>Have you got a miraculous new compression algorithm to share ? :)

How about BCG (Binary Coded Gcode 😁)? Bit seven in letter set indicates trailing binary coded number. And you just need a small parser between your GCODE interpreter and your EEPROM.
  
What the hell are you going to store? Plenty of space, even for checksums header, tail and so on. 😕 

Perhaps just count before flaming? Don't find problems, find solutions. 👍

**Comment by CONSULitAS** [comment]
> How about BCG (Binary Coded Gcode 😁)? Bit seven in letter set indicates trailing binary coded number. And you just need a small parser between your GCODE interpreter and your EEPROM.

BTW: You only need 5 bit to encode a case insensitive letter. So bits 6+7 could tell the data type:
* 00: plain ascii - for text 
* 01: 1 byte - number
* 10: 2 byte integer
* 11: float

Perhaps this is the __miraculous new compression algorithm__ you are looking for?

I think you will end with about 200 bytes if well done.

And perhaps we will really need it if @Roxy-3DPrintBoard tries to save huge ABL data sets. 💭

**Comment by CONSULitAS** [comment]
>We could even skip spaces to save some bytes...

```
wc -c MarlinConfig_2016-04-16_nospace.gcode 
     282  MarlinConfig_2016-04-16_nospace.gcode
```

Saved 42 bytes. Now below 300 bytes.

Perhaps with BCG 150 bytes.

Any problems with EEPROM space left? Lets solve them!

**Comment by CONSULitAS** [comment]
I think i just reinvented the wheel: 🎱 😁 

http://reprap.org/wiki/Packed_Binary_GCode

**Comment by jbrazio** [comment]
Gcode suffers exactly from the same issue has xml, in order to be human readable the size is not optimized. The natural evolution will push 3D printers into a more "dense" protocol sooner or later as serial speeds will not increase.

In fact I believe for the next generation of Marlin we should support multiple input protocols: GCode, Repetier, PBG, whatever. It will only be a question of having an object extending the core parser.

But internally, i.e. planner buffers, we should switch to a new dense format as using the same buffer space we could potentially store more commands.

**Comment by bobc** [comment]
The config file for RepRapFirmware is 2.7K, the Smoothie one is 27K
although they use a more wordy format.

It will be very easy to blow the 4K size once you start adding stuff. Sure,
there are easy ways to compress data. But you have to think *compact* if
you only have 4K. You certainly can't create a format suitable for an SD
card and expect the same to work on EEPROM. The other way round, yes.

Config files always grow bigger than you expect. How much space will the
high-res data bed levelling take ?

**Comment by jbrazio** [comment]
But.. the config file overall size has practically nothing to do with EEPROM.. unless we want to remove all compile time directives and use only runtime directives.

**Comment by bobc** [comment]
Of course I realise that. Those are the sizes of the config file they store on the SD card.

**Comment by gddeen** [comment]
Is there a tool to print out a reference of the PINs (after compiled), by name, by pin number?

The new dev is moving everything to new places, but I just really want to know if the PINs are the
same. (I.E. I selected a MKS board, but the servo1 pin had changed. I also override  Z_MIN_PROBE_PIN to 15.)

That's probably the bottom line is if all my PINouts still match, pick up that build and try it.

**Comment by gddeen** [comment]
Persisting objects for EEPROM should really persist state data, not necessarily 'code' to restore the data.
However, if persistence occurs over WiFi, Ethernet, or some OctoPrint like, it can start having bells and whistles. (IE editable by smart users.)

After awhile if unneeded, EEPROM will be eliminated because no-one wants it.  My printer has an SD card.  If it persisted everything there I would use that to my advantage possibly.

At some point EEPROM won't be used because it is just too slow and too small.

**Comment by thinkyhead** [comment]
A thought. We could provide a function to help with upgrades where the EEPROM layout might change. It would simply save the output of `M503 S0` to the SD Card, then you could just run that file to restore it after the version upgrade. Currently you would have to copy it from the console and save it yourself.

**Comment by thinkyhead** [comment]
@gddeen I think you're probably correct about that with some boards. For users who don't use SD printing, EEPROM will still have value for a while, though.

**Comment by clexpert** [comment]
@thinkyhead @jbrazio @Roxy-3DPrintBoard Speaking of c++ singletons, you asked me to write Filament load, unload and change using singletons. It gives me two ideas of how to do it: either use planner (`line_to_destination`) or use g-codes (`enqueue("G28")`). Former seems code-prettier, the latter is universal (cartesian and delta, `G1` does move and asks for no config). 
But there is still a big problem. When written in `ultralcd` or `Marlin_main`, is has access to all movement, temperature and extruder features. With the idea of "code" writing in separate file, things like `line_to_destination`, `extruder_current` or `equeue g-code` would have to be external visible.
In would be nice to write `filament.unload`, `filament.change` or even `filament.hasRunOut`.
Which strategy would you choose and how to solve dependencies?

**Comment by AnHardt** [comment]
How about `prepare_move()`? Using `enqueue()` makes a huge overhead - especially if you have to build the string from varying parameters.

**Comment by Roxy-3DPrintBoard** [comment]
I am a little nervous about using GCodes to accomplish the work.   We should get everybody to weigh in with their thoughts but if there is a support routine you need to accomplish the work it probably makes sense to declare it in Marlin.h so it can be used where appropriate.   

And conversely, looking to the future...  Any useful member functions that you define should be available by any module that #includes your .h file.  (Probably we don't want to add those declarations to Marlin.h    Declaring stuff you need in Marlin.h is just a necessary evil because the code hasn't been modulized yet.)

Please help me understand something.   I'm a little bit hung up on a technicality.   A 'singleton' by definition has only one instance and it is global and pre-initialized at program startup.  Would your 'singleton' handle multiple extruders (and therefor multiple filaments)?    Or would you actually have two (or three or four) of these singleton's with each one handling one extruder's filament?

It doesn't really matter as long as it is done cleanly.   But I'm trying to figure out if this is literally a singleton.

**Comment by gddeen** [comment]
Currently, line_to_destination, etc should be actions (methods) of a "planner" or "movement" object.
So, a filament object has actions like unload and change and at the start uses the old
non objectized global functions.  When the planner object is created, and the "line to destination"
method is created, all function use will be removed and objects used in their place.

Codes should not be used because they are not the "peer" objects.  "G1 X0" might have different
implementations, etc and don't use the object created to perform the particular work exactly.

But, I dunno. just 2 cents.

**Comment by gddeen** [comment]
@clexpert Sorry I responded... I'm not a Marlin developer, just a developer that plays with Marlin.

**Comment by jbrazio** [comment]
@clexpert this is a good question.

First of all I would prefer to programmatically plan the movements instead of using `G-Code`, mainly because specific G-Code can change the syntax or be added new features that can "chain-brake" dependencies. One of the objectives of going to objects is to minimize the risk of braking features when specific parts of the code are being worked on.

The problem with this feature is where to consider it will be built from, i.e. should we have a standalone object to handle this or would it be a method/[or] extension of the `extruder` object, which does not exist yet ? IMO the final resting place for this will be inside the `extruder` object but for now..

I would suggest to proceed as you proposed, create the object `Filament`, and define the **public** methods: `unload()`, `load()`, `change()` and `hasRunOut()`.

In `Marlin.h` we declare `extern Filament filament;`, in `Marlin_main.cpp` we declare something like `Filament filament = Filament();` and this will allow all the public methods to be accessible to Marlin as long as they `#include "Marlin.h".

@Roxy-3DPrintBoard 
> Would your 'singleton' handle multiple extruders (and therefor multiple filaments)? Or would you actually have two (or three or four) of these singleton's with each one handling one extruder's filament?

I would say this new `Filament` object would, for now, have a public method `setTargetExtruder()` so we can shift extruders, in the future if we end up having one `Extruder` object per extruder, then it will be easier to merge in the code as we just have to override the `Filament::constructor().

**Comment by Sniffle** [comment]
I would add to that list of variables for filament diameter() as well.

**Comment by Roxy-3DPrintBoard** [comment]
> I would add to that list of variables for filament diameter() as well.

I'm not a C++ expert.   But just to help me understand better, diameter() would be a member function and not a variable, right?   And if I was writing the code, I would consider having it over loaded.   If called with no parameters it would return the current diameter of the filament.  If called with a parameter, it would set the diameter to that number and then return that same number....    But I'm sure clexpert doesn't need any back seat drivers!   :)

**Comment by Sniffle** [comment]
i quite probably used incorrect wording, but adding diameter to the
filament object allows it to be tied to the diameter sensor and advance, or
just be statically set

On Mon, May 23, 2016 at 4:22 PM, Roxy-3DPrintBoard <notifications@github.com

**Comment by gddeen** [comment]
@thinkyhead I was thinking fast, 32bit large RTOS with a hugh footprint hovering at the embedded/standalone computer. Things that can rotate the bed level and even show
visually the entire gcode.

**Comment by jbrazio** [comment]
@gddeen are you positive you replied in the correct issue ?

**Comment by clexpert** [comment]
Thank you everybody for help, code is getting clearer and soon it'll be ready I hope. 
But last questinon: @AnHardt recommended awesome global function `prepare_move` which really does what I need. But it calls `line_to_destination` which moves axes to `destination`, but I (from other file) have no access to that variable, beacuse it is `static`. Am I missing any function in `Marlin.h` which sets the destination?

**Comment by Roxy-3DPrintBoard** [comment]
>  but I (from other file) have no access to that variable, because it is static. Am I missing any function in Marlin.h which sets the destination?

If you need access to some global variable, you can add a declaration to Marlin.h

I'm not exactly clear on what the problem is, but if it is because destination[] is not declared within your scope, you can add:
~~~~
extern static float destination[NUM_AXIS]; 
~~~~
to Marlin.h so you can compile and keep moving forward.

**Comment by Blue-Marlin** [comment]
@Roxy-3DPrintBoard 
You entirely missed the sense of an `static` variable declaration. When not declared in a function it means - for this .cpp file only.

If `static` is omittet your `extern static float destination[NUM_AXIS]; ` in Marlin.h should work.

However, i'd prefer a version with a function.
Maybe like: (In Marlin.cpp)
```cpp
void set_destination(float *source[NUM_AXES]) {  // or something like that
  memcpy(destination, source, sizeof(destination)); 
}
void set_destination(float x, float y, float z, float e) {
  destination[X_AXIS] = x;
  destination[Y_AXIS] = y;
  destination[Z_AXIS] = z;
  destination[E_AXIS] = e;
}
```
and the matching declarations in Marlin.h

But obviously `current_position[]` is already global, so it does not matter any more.

**Comment by Roxy-3DPrintBoard** [comment]
OK.  I did not know that static had a different meaning depending upon if it is used inside a function or not.   You learn something every day!

**Comment by clexpert** [comment]
@Roxy-3DPrintBoard As I learned today, `static` outside class means sth as `private` in other languages.
@Blue-Marlin Thank you for advice, I thing Marlin lacks some essential functions like this `set_destination`. So you agree with making this new global function? Looks like someone should look on support for future modules a handle shared code. :smile:

**Comment by gddeen** [comment]
@Roxy-3DPrintBoard 'Diameter' is an attribute of some object. So that would naturally be a variable somewhere.  It can be within an object such as Filament. It can be public or private (only filament can see it.) And then usually a member function is written like 'getDiameter()' I guess because programmers
used to be graded on lines of code in a release.

As a side note the smartest programmers may delete a lot more that they write.

@jbrazio I dunno if I'm in any correct forum. I see a question somewhere, I jump in and help.
I'll be quieter...

**Comment by jbrazio** [comment]
@Blue-Marlin :+1: 

@gddeen nvm I saw you're replying to an older post.

@clexpert another use of `static` when inside a class means that the method/var is from the class and not from the instance thus a static class var is "shared" among all the class instances. In a weird way you may say it's equivalent to `private` in the sense that all instances of the same class can access each other's private methods.

**Comment by thinkyhead** [comment]
> no objects with public vars unless with a very good reason, always use getters/setters.

Academically, I agree, and pedantically, yeah. But realistically I see no practical reason to be religious about getters/setters for simple values that are unlikely to change (or become something more complex) behind the scenes. If you must write getters and setters for simple values, strive to make them `FORCE_INLINE` whenever possible.

**Comment by thinkyhead** [comment]
> `static` outside `class` means sth as `private` in other languages

Well, sort of. A `private` method might still be made visible to other translation units (specified by class) through the `friend` keyword. You don't have any such option with `static` C-style functions.

Stack Overflow defines `static` like so: "In C, a `static` function is not visible outside of its translation unit, which is the object file it is compiled into. In other words, making a function `static` limits its scope."

So in that sense, the function is "private" — only visible within its own translation unit — and that is the opposite of the default. A C-style function _without_ `static` can also be declared elsewhere (usually the header for the .c/.cpp file it appears in) and used in other translation units. The compiler makes an external reference so the linker can find it and link it. Declaring a C (global) function in a header file is the same as declaring it `extern`. The linker is left to resolve it later. (You can leave off the keyword `extern` for global _functions_, but you must use `extern` for global _variables_.)

**Comment by jbrazio** [comment]
Devs do not have to care about any object's implementation if interacting only trough the public interfaces, having class vars which can be changed from the exterior without control makes it loose state which in turn breaks the concept.

But yeah you're right, sometimes the overhead is an overkill, in the end of the day all needs a bit of salt.
